util/winutil/restartmgr_windows.go

```diff
@@ -0,0 +1,830 @@
+	SessionInfo SessionID
+	// CommandLineInfo contains the command line information necessary for restarting.
+	CommandLineInfo CommandLineInfo
+	// UserSID contains the stringified SID of the existing process's token's user.
+	UserSID string
+	// handle specifies the Win32 HANDLE associated with the existing process.
+	// When non-zero, it includes access rights for querying, terminating, and synchronizing.
+	handle windows.Handle
+	// hasExitCode is true when the exitCode field is valid.
+	hasExitCode bool
+	// exitCode contains exit code returned by this RestartableProcess once
+	// its termination has been recorded by (RestartableProcesses).Terminate.
+	// It is only valid when hasExitCode == true.
+	exitCode uint32
+}
+
+func (rp *RestartableProcess) Close() error {
+	if rp.handle == 0 {
+		return nil
+	}
+	if err := windows.CloseHandle(rp.handle); err != nil {
+		return err
+	}
+	rp.handle = 0
+	runtime.SetFinalizer(rp, nil)
+	return nil
+}
+
+// RestartableProcesses is a map of PID to *RestartableProcess instance.
+type RestartableProcesses map[uint32]*RestartableProcess
+
+// NewRestartableProcesses instantiates a new RestartableProcesses.
+func NewRestartableProcesses() RestartableProcesses {
+	return make(RestartableProcesses)
+}
+
+// Add inserts rp into rps.
+func (rps RestartableProcesses) Add(rp *RestartableProcess) {
+	if rp != nil {
+		rps[rp.Process.PID] = rp
+	}
+}
+
+// Delete removes rp from rps.
+func (rps RestartableProcesses) Delete(rp *RestartableProcess) {
+	if rp != nil {
+		delete(rps, rp.Process.PID)
+	}
+}
+
+// Close invokes (*RestartableProcess).Close on every value in rps, and then
+// clears rps.
+func (rps RestartableProcesses) Close() error {
+	for _, v := range rps {
+		v.Close()
+	}
+	clear(rps)
+	return nil
+}
+
+// _MAXIMUM_WAIT_OBJECTS is the Win32 constant for the maximum number of
+// handles that a call to WaitForMultipleObjects may receive at once.
+const _MAXIMUM_WAIT_OBJECTS = 64
+
+// Terminate forcibly terminates all processes in rps using exitCode, and then
+// waits for their process handles to signal, up to timeout.
+func (rps RestartableProcesses) Terminate(logf logger.Logf, exitCode uint32, timeout time.Duration) error {
+	if len(rps) == 0 {
+		return nil
+	}
+
+	millis, err := wingoes.DurationToTimeoutMilliseconds(timeout)
+	if err != nil {
+		return err
+	}
+
+	errs := []error{}
+	procs := make([]*RestartableProcess, 0, len(rps))
+	handles := make([]windows.Handle, 0, len(rps))
+	for _, v := range rps {
+		if err := windows.TerminateProcess(v.handle, exitCode); err != nil {
+			if err == windows.ERROR_ACCESS_DENIED {
+				// If v terminated before we attempted to terminate, we'll receive
+				// ERROR_ACCESS_DENIED, which is not really an error worth reporting in
+				// our use case. Just obtain the exit code and then close the process.
+				if err := windows.GetExitCodeProcess(v.handle, &v.exitCode); err != nil {
+					logf("GetExitCodeProcess failed: %v", err)
```

Comment:
Should this `err` be added to `errs`? Like `errs = append(errs, &terminationError{rp: v, err: fmt.Errorf("GetExitCodeProcess: %w", err)})`

Reply:
I'm tempted to say no, because the process *has* terminated, we just failed to get its error code. If we did return an error, the net effect would be that it would cause the installer to abort and rollback, which is not really our desired result.